physical/gcs/gcs.go

```diff
@@ -214,3 +257,271 @@ func (g *GCSBackend) List(prefix string) ([]string, error) {
 
 	return keys, nil
 }
+
+// GCSLock implements a lock using an GCS client.
+type GCSLock struct {
+	backend  *GCSBackend
+	value    string
+	key      string
+	identity string
+	held     bool
+	lock     sync.Mutex
+	// Allow modifying the Lock durations for ease of unit testing.
+	renewInterval      time.Duration
+	ttl                time.Duration
+	watchRetryInterval time.Duration
+}
+
+// LockWith is used for mutual exclusion based on the given key.
+func (g *GCSBackend) LockWith(key, value string) (physical.Lock, error) {
+	identity, err := uuid.GenerateUUID()
+	if err != nil {
+		return nil, err
+	}
+	return &GCSLock{
+		backend:            g,
+		key:                key,
+		value:              value,
+		identity:           identity,
+		renewInterval:      GCSLockRenewInterval,
+		ttl:                GCSLockTTL,
+		watchRetryInterval: GCSWatchRetryInterval,
+	}, nil
+}
+
+func (g *GCSBackend) HAEnabled() bool {
+	return g.haEnabled
+}
+
+// Lock tries to acquire the lock by repeatedly trying to create
+// a record in the GCS. It will block until either the
+// stop channel is closed or the lock could be acquired successfully.
+// The returned channel will be closed once the lock is deleted or
+// changed in the GCS.
+func (l *GCSLock) Lock(stopCh <-chan struct{}) (doneCh <-chan struct{}, retErr error) {
```

Comment:
We should [publish metrics](https://github.com/hashicorp/vault/blob/5b4abc09ddd82be99e05264a91b06bc18ab1447a/physical/spanner/spanner_ha.go#L42-L51) for all the locking events.